<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN"
    "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<meta name="generator" content="AsciiDoc 8.2.6" />
<style type="text/css">
/* Debug borders */
p, li, dt, dd, div, pre, h1, h2, h3, h4, h5, h6 {
/*
  border: 1px solid red;
*/
}

body {
  margin: 1em 5% 1em 5%;
}

a {
  color: blue;
  text-decoration: underline;
}
a:visited {
  color: fuchsia;
}

em {
  font-style: italic;
  color: navy;
}

strong {
  font-weight: bold;
  color: #083194;
}

tt {
  color: navy;
}

h1, h2, h3, h4, h5, h6 {
  color: #527bbd;
  font-family: sans-serif;
  margin-top: 1.2em;
  margin-bottom: 0.5em;
  line-height: 1.3;
}

h1, h2, h3 {
  border-bottom: 2px solid silver;
}
h2 {
  padding-top: 0.5em;
}
h3 {
  float: left;
}
h3 + * {
  clear: left;
}

div.sectionbody {
  font-family: serif;
  margin-left: 0;
}

hr {
  border: 1px solid silver;
}

p {
  margin-top: 0.5em;
  margin-bottom: 0.5em;
}

ul, ol, li > p {
  margin-top: 0;
}

pre {
  padding: 0;
  margin: 0;
}

span#author {
  color: #527bbd;
  font-family: sans-serif;
  font-weight: bold;
  font-size: 1.1em;
}
span#email {
}
span#revision {
  font-family: sans-serif;
}

div#footer {
  font-family: sans-serif;
  font-size: small;
  border-top: 2px solid silver;
  padding-top: 0.5em;
  margin-top: 4.0em;
}
div#footer-text {
  float: left;
  padding-bottom: 0.5em;
}
div#footer-badges {
  float: right;
  padding-bottom: 0.5em;
}

div#preamble,
div.tableblock, div.imageblock, div.exampleblock, div.verseblock,
div.quoteblock, div.literalblock, div.listingblock, div.sidebarblock,
div.admonitionblock {
  margin-right: 10%;
  margin-top: 1.5em;
  margin-bottom: 1.5em;
}
div.admonitionblock {
  margin-top: 2.5em;
  margin-bottom: 2.5em;
}

div.content { /* Block element content. */
  padding: 0;
}

/* Block element titles. */
div.title, caption.title {
  color: #527bbd;
  font-family: sans-serif;
  font-weight: bold;
  text-align: left;
  margin-top: 1.0em;
  margin-bottom: 0.5em;
}
div.title + * {
  margin-top: 0;
}

td div.title:first-child {
  margin-top: 0.0em;
}
div.content div.title:first-child {
  margin-top: 0.0em;
}
div.content + div.title {
  margin-top: 0.0em;
}

div.sidebarblock > div.content {
  background: #ffffee;
  border: 1px solid silver;
  padding: 0.5em;
}

div.listingblock {
  margin-right: 0%;
}
div.listingblock > div.content {
  border: 1px solid silver;
  background: #f4f4f4;
  padding: 0.5em;
}

div.quoteblock > div.content {
  padding-left: 2.0em;
}

div.attribution {
  text-align: right;
}
div.verseblock + div.attribution {
  text-align: left;
}

div.admonitionblock .icon {
  vertical-align: top;
  font-size: 1.1em;
  font-weight: bold;
  text-decoration: underline;
  color: #527bbd;
  padding-right: 0.5em;
}
div.admonitionblock td.content {
  padding-left: 0.5em;
  border-left: 2px solid silver;
}

div.exampleblock > div.content {
  border-left: 2px solid silver;
  padding: 0.5em;
}

div.verseblock div.content {
  white-space: pre;
}

div.imageblock div.content { padding-left: 0; }
div.imageblock img { border: 1px solid silver; }
span.image img { border-style: none; }

dl {
  margin-top: 0.8em;
  margin-bottom: 0.8em;
}
dt {
  margin-top: 0.5em;
  margin-bottom: 0;
  font-style: normal;
}
dd > *:first-child {
  margin-top: 0.1em;
}

ul, ol {
    list-style-position: outside;
}
div.olist > ol {
  list-style-type: decimal;
}
div.olist2 > ol {
  list-style-type: lower-alpha;
}

div.tableblock > table {
  border: 3px solid #527bbd;
}
thead {
  font-family: sans-serif;
  font-weight: bold;
}
tfoot {
  font-weight: bold;
}

div.hlist {
  margin-top: 0.8em;
  margin-bottom: 0.8em;
}
div.hlist td {
  padding-bottom: 15px;
}
td.hlist1 {
  vertical-align: top;
  font-style: normal;
  padding-right: 0.8em;
}
td.hlist2 {
  vertical-align: top;
}

@media print {
  div#footer-badges { display: none; }
}

div#toctitle {
  color: #527bbd;
  font-family: sans-serif;
  font-size: 1.1em;
  font-weight: bold;
  margin-top: 1.0em;
  margin-bottom: 0.1em;
}

div.toclevel1, div.toclevel2, div.toclevel3, div.toclevel4 {
  margin-top: 0;
  margin-bottom: 0;
}
div.toclevel2 {
  margin-left: 2em;
  font-size: 0.9em;
}
div.toclevel3 {
  margin-left: 4em;
  font-size: 0.9em;
}
div.toclevel4 {
  margin-left: 6em;
  font-size: 0.9em;
}
/* Workarounds for IE6's broken and incomplete CSS2. */

div.sidebar-content {
  background: #ffffee;
  border: 1px solid silver;
  padding: 0.5em;
}
div.sidebar-title, div.image-title {
  color: #527bbd;
  font-family: sans-serif;
  font-weight: bold;
  margin-top: 0.0em;
  margin-bottom: 0.5em;
}

div.listingblock div.content {
  border: 1px solid silver;
  background: #f4f4f4;
  padding: 0.5em;
}

div.quoteblock-content {
  padding-left: 2.0em;
}

div.exampleblock-content {
  border-left: 2px solid silver;
  padding-left: 0.5em;
}

/* IE6 sets dynamically generated links as visited. */
div#toc a:visited { color: blue; }

/* Because IE6 child selector is broken. */
div.olist2 ol {
  list-style-type: lower-alpha;
}
div.olist2 div.olist ol {
  list-style-type: decimal;
}
</style>
<title>Trinity's In silico Read Normalization</title>
</head>
<body>
<div id="header">
<h1>Trinity's In silico Read Normalization</h1>
</div>
<div id="preamble">
<div class="sectionbody">
<div class="para"><p><a href="index.html">Trinity-home</a></p></div>
<div class="para"><p>Large RNA-Seq data sets, such as those exceeding 300M pairs, are best suited for in silico normalization prior to running Trinity, in order to reduce memory requirements and greatly improve upon runtimes.</p></div>
<div class="admonitionblock">
<table><tr>
<td class="icon">
<div class="title">Note</div>
</td>
<td class="content">Before running the normalization, be sure that in the case of paired reads, the left read names end with suffix <em>/1</em> and the right read names end with <em>/2</em>.</td>
</tr></table>
</div>
<div class="para"><p>To normalize your data set, run the included:</p></div>
<div class="literalblock">
<div class="content">
<pre><tt>TRINITY_RNASEQ_ROOT/util/normalize_by_kmer_coverage.pl</tt></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><tt>###############################################################################
#
# Required:
#
#  --seqType &lt;string&gt;      :type of reads: ( 'fq' or 'fa')
#  --JM &lt;string&gt;            :(Jellyfish Memory) number of GB of system memory to use for
#                            k-mer counting by jellyfish  (eg. 10G) *include the 'G' char
#
#  --max_cov &lt;int&gt;         :targeted maximum coverage for reads.
#
#
#  If paired reads:
#      --left  &lt;string&gt;    :left reads
#      --right &lt;string&gt;    :right reads
#
#  Or, if unpaired reads:
#      --single &lt;string&gt;   :single reads
#
#  Or, if you have read collections in different files you can use 'list' files, where each line in a list
#  file is the full path to an input file.  This saves you the time of combining them just so you can pass
#  a single file for each direction.
#      --left_list  &lt;string&gt; :left reads, one file path per line
#      --right_list &lt;string&gt; :right reads, one file path per line
#
####################################
##  Misc:  #########################
#
#  --pairs_together                :process paired reads by averaging stats between pairs and retaining linking info.
#
#  --SS_lib_type &lt;string&gt;          :Strand-specific RNA-Seq read orientation.
#                                   if paired: RF or FR,
#                                   if single: F or R.   (dUTP method = RF)
#                                   See web documentation.
#  --output &lt;string&gt;               :name of directory for output (will be
#                                   created if it doesn't already exist)
#                                   default( "normalized_reads" )
#
#  --JELLY_CPU &lt;int&gt;                     :number of threads for Jellyfish to use (default: 2)
#  --PARALLEL_STATS                :generate read stats in parallel for paired reads (Figure 2X Inchworm memory requirement)
#
#  --KMER_SIZE &lt;int&gt;               :default 25
#
#  --min_kmer_cov &lt;int&gt;            :minimum kmer coverage for catalog construction (default: 1)
#
#  --max_pct_stdev &lt;int&gt;           :maximum pct of mean for stdev of kmer coverage across read (default: 100)
#
###############################################################################</tt></pre>
</div></div>
<div class="para"><p>This should be run on a machine that has a suitably large amount of RAM (typically hundreds of GB of RAM).
The command-line options are quite similar to those used by Trinity.pl.</p></div>
<div class="para"><p>There are several potential invocations of the normalization process, depending on your interests.</p></div>
<div class="para"><p>Ideally, the process would be run as follows:</p></div>
<div class="literalblock">
<div class="content">
<pre><tt>TRINITY_RNASEQ_ROOT/util/normalize_by_kmer_coverage.pl --seqType fq --JM 100G --max_cov 30 --left left.fq --right right.fq --pairs_together --PARALLEL_STATS --JELLY_CPU 10</tt></pre>
</div></div>
<div class="para"><p>To roughly halve memory requirements but take almost twice as long to run, remove the <em>&#8212;PARALLEL_STATS</em> parameter.  In this case, the left.fq and right.fq files will be examined separately instead of in parallel.</p></div>
<div class="para"><p>Also, to greatly lessen memory requirements, include the option <em>&#8212;min_kmer_cov 2</em>, in which case no uniquely occurring kmer will be assayed.</p></div>
</div>
</div>
<h2 id="_results">Results</h2>
<div class="sectionbody">
<div class="para"><p>Normalized versions of the reads will be written, with file names corresponding to the original files but with an added extension to indicate the targeted maximum coverage levels.</p></div>
</div>
<h2 id="_sample_data">Sample data</h2>
<div class="sectionbody">
<div class="para"><p>You can run the normalization process on the sample data provided at:</p></div>
<div class="literalblock">
<div class="content">
<pre><tt>TRINITY_RNASEQ_ROOT/sample_data/test_Trinity_Assembly</tt></pre>
</div></div>
<div class="para"><p>by running</p></div>
<div class="literalblock">
<div class="content">
<pre><tt>__run_read_normalization_pairs_together_fastq.sh</tt></pre>
</div></div>
<div class="para"><p>(note there are several such <em>__run_read_normalization&#8230;</em> scripts to demonstrate alternate scenarios for execution.)</p></div>
</div>
<div id="footer">
<div id="footer-text">
Last updated 2012-12-01 14:56:21 EDT
</div>
</div>
</body>
</html>
